### ISSUE TITLE
Suggestion: global addContextProps hook

### ISSUE THREAD

**Issue by gryphonmyers** [brillout/vite-plugin-ssr#26]
I have a need for global data that is available to all pages. I am aware that this is possible within the server entrypoint of the application, however that is not a perfect solution in all cases for the following reasons:

- The server file entrypoint is not a module context, so does not have access to env variables via `import.meta.env`. One could maybe run `dotenv.config` to get around this, but it feels inconsistent and hacky to be accessing env vars in two different ways
- Since the server entrypoint is not processed by Vite, it complains when attempting to import files using ES module syntax. This can lead to inconsistencies, confusion around which files can / are supposed to be used where.
- Data populated in the server entrypoint is not available in prerender hooks. This means any logic used here to populate the initial contextProps needs to be replicated in prerender hooks.

I am proposing the addition of API that would allow for global additions to context props covering all of these areas. I don't have a specific API in mind yet, just opening the discussion as I see a need here.

**Comment by brillout** [comment]
Makes sense. I'll think about all that.

First reaction: maybe something like https://github.com/axe-me/vite-plugin-node#readme is the solution for 1. and 2. But I'm also thinking that something built in `vite-plugin-ssr` may be better; it can stay optional to stay true to "do-one-thing-do-it-well".

First reaction about 3.: maybe we could allow `_default.server.js` to have a `addContextProps()` but that may lead to confusions. I'll think about this.

> I have a need for global data that is available to all pages

I'm curious; what is your concrete use case?

**Comment by gryphonmyers** [comment]
> Makes sense. I'll think about all that.
> 
> First reaction: maybe something like https://github.com/axe-me/vite-plugin-node#readme is the solution for 1. and 2. But I'm also thinking that something built in `vite-plugin-ssr` may be better; it can stay optional to stay true to "do-one-thing-do-it-well".
> 
> First reaction about 3.: maybe we could allow `_default.server.js` to have a `addContextProps()` but that may lead to confusions. I'll think about this.
> 
> > I have a need for global data that is available to all pages
> 
> I'm curious; what is your concrete use case?

We typically have some amount of data that sits at the application level, which would be tedious and error-prone to resolve independently in each page. This tends to be things like the locales supported for i18n, dictionary of translation strings, application metadata (title, color, JSON-LD, etc).

I would also see opportunity to populate data that is unique to a particular page, but using global parameterized functionality. For instance,  say you want every page to fetch some "page" data describing its contents from your API based on the current path. This page data will always be fetched in the same way using the path currently being rendered, so it would keep the code DRY if you could write it once in a global hook of some kind.

As for the prerender hook, it would mostly just be nice not to have to start from scratch on populating data there. For instance, the locales array I mentioned before - having access to this in the prerender hook is required for generating out all the locale-specific paths, so it would be great if that hook also had access to this part of the context.

Hopefully that helps illustrate.

**Comment by gryphonmyers** [comment]
Thinking more about this, I might be describing two separate hooks. One that executes only one time at initialization (either at server start or prerender process initilization), and one that executes per request. The one that executes per request kind of doesn't make sense in a prerender hook because the prerender hook itself is the thing determining which "requests" should be made. However, it might be nice if the prerender hook had the ability to itself call the per-request hook in order to resolve data from it that it could then pass into the context.

**Comment by gryphonmyers** [comment]
> First reaction about 3.: maybe we could allow _default.page.server.js to have a addContextProps() (edit: in addition to the page's addContextProps(); currently, the page's addContextProps() overrides the one defined in _default.page.server.js) but that may lead to confusions. I'll think about this.

I was thinking about that too - overriding honestly probably makes more intuitive sense. What about just making a totally new hook like `initializeContextProps`. Then the default could use this hook, while the pages use the other one and there would be no ambiguity. Logically then the pages would also be able to override this hook, which might be pointless or maybe there's some use case where that makes sense. In my mind, when the request comes in it would call initializeContextProps for the default, then for any pages that implement it, then repeat for addContextProps.

I suppose the hook I'm describing now would be the "per request" one from my earlier comment. I'm honestly not sure where the process initialization hook would go. What if it was just an option that could be passed into the plugin? Then you'd be within the vite config and firmly in ES module territory, which would sort out (I think) problems 1 and 2.

**Comment by brillout** [comment]
Just for the sake of the discussion; would something like this work for you?

```js
// pages/initializeContextProps.js

export { initializeContextProps }

async function initializeContextProps({ contextProps }) {
   // A function that does whatever I want and that I use wherever I want.
}
```

```js
// pages/_default.page.server.js

import { initializeContextProps } from './initializeContextProps'

export { addContextProps }

async function addContextProps({ contextProps }) {
  initialContextProps = await initializeContextProps(contextProps)
}
```

```js
// pages/product.page.server.js

import { initializeContextProps } from './initializeContextProps'

export { addContextProps }

async function addContextProps({ contextProps }) {
  // We overrides the default thus we re-use `initializeContextProps()`.
  initialContextProps = await initializeContextProps(contextProps)

  // More stuff
  const contextPropsAddendum = /*...*/

  return {...initialContextProps, ...contextPropsAddendum}
}
```

**Comment by gryphonmyers** [comment]
Hey not a bad suggestion - let me try that and see if there's any reason why that would be insufficient.

**Comment by brillout** [comment]
One cosmetic imperfection with that solution: if you use `initializeContextProps` at the server integration point, then you'll have to write it with CJS.

You're right, we should dig deeper into enabling the server code to be processed by Vite. In the meantime, `nodemon` + CJS + a `env.js` is the way to go. Not pretty, but works.

```js
// env.js

const env = {}

if( typeof window !== 'undefined' ){
  window.env = env
} else {
  // Node.js
  global.env = env
}
```
```js
import './env' // for page code
require('./env') // for server code

env.someGlobalVariable = 'I am available to everyone';
```

**Comment by brillout** [comment]
> enabling the server code to be processed by Vite

I'm thinking maybe simply a `vite-node`, similarly to `ts-node`. I'm going to play around that idea.

**Comment by gryphonmyers** [comment]
OK sounds good let me know what you settle on and if you'd like me to try it out to see if it satisfies my use cases.

**Comment by gryphonmyers** [comment]
Here's one other thought for a possible optimization here: if we have a concept of global context props, we can exclude them from *.pageProps.json files and have a single json file that only gets fetched once (likely from DOM cached SSR snapshot). This would avoid bloating route-specific *.pageProps.json with duplicate data.

**Comment by brillout** [comment]
> OK sounds good let me know what you settle on and if you'd like me to try it out to see if it satisfies my use cases.

There something quite neat WIP.

> Here's one other thought for a possible optimization here: if we have a concept of global context props, we can exclude them from *.pageProps.json files and have a single json file that only gets fetched once (likely from DOM cached SSR snapshot). This would avoid bloating route-specific *.pageProps.json with duplicate data.

When deploying to a static host, we need one static file per page. Otherwise all data for all pages would need to be loaded which would slow down the rendering of the page.

**Comment by brillout** [comment]
Ah, you mean data that is needed by all pages. Such data is usually static and defined in source code; or do you have an example of semi-dynamic data that needs to be computed dynamically but that is common to all pages? Doesn't seem to be that common?

**Comment by gryphonmyers** [comment]
> Ah, you mean data that is needed by all pages. Such data is usually static and defined in source code; or do you have an example of semi-dynamic data that needs to be computed dynamically but that is common to all pages? Doesn't seem to be that common?

In our case this data is not defined in source code, but comes through via API calls to our headless CMS. This is application-specific configuration I mentioned before, things like locales, application metadata, etc.

**Comment by gryphonmyers** [comment]
It would be fetched exactly once during compilation and thus the same for all routes

**Comment by brillout** [comment]
I see. What you could do is to inject that data in HTML and, since you are doing client-side routing, the data will be loaded only once.

**Comment by gryphonmyers** [comment]
That's exactly what I'm doing now. It feels a bit dirty though serializing that data and storing it in a separate window namespace, when this is data that the plugin knows about.

**Comment by brillout** [comment]
Or generate a `globalContext.js` file. You can then cache it as you wish, e.g. `globalContext.123iha9u21.js` where `123iha9u21` is a content hash so that you can cache the URL path forever.

**Comment by brillout** [comment]
The current recommendation is to simulate a global hook by implementing some function, [e.g. `initializeContextProps()`](https://github.com/brillout/vite-plugin-ssr/issues/26#issuecomment-811763190), that you then call everywhere.

**Comment by gryphonmyers** [comment]
> > enabling the server code to be processed by Vite
> 
> I'm thinking maybe simply a `vite-node`, similarly to `ts-node`. I'm going to play around that idea.

Did your investigation turn up any useful guidance toward getting the server entrypoint to be processed by Vite?

**Comment by brillout** [comment]
I got there halfway but it's not usable yet. I can make the work-in-progress public if you want.

**Comment by gryphonmyers** [comment]
IDK just, after spending some time with the solution described above, the ergonomics are _ok_ but still a little WET for data that is truly global. That kind of data really seems like it should be passed into the server entrypoint but the CJS requirement is kind of a show-stopper. 

Looking at this problem through the lens of recent discussions in other threads, the suggestion of event-based hooks seems like it could be intuitive here. 

_default.page.server.js

javascript```

onInitializeContext({ pageContext }) {
  return {
    ...pageContext, // runs first, once for each route/request. contains built-in properties like url and whatnot
    myGlobalData
  }
}

onBeforeRender({ pageContext }) {
  return {
    ...pageContext, // runs second, once for each route/request. contains result from onInitializeContext
    pageSpecificData
  }
}

```

This would avoid the need to comprehensively apply "myGlobalData" in all page files, as long as `onInitializeContext` is not overridden. If you override onInitializeContext in a specific page file, it would be your responsibility to bring along any needed data that was applied in the default hook. The point though is you usually don't have to worry about it, as opposed to now where you _do_ have to worry about it, in every single page file. If you forget to include some necessary piece of global state in a new page, you get errors.

**Comment by brillout** [comment]
@gryphonmyers

That makes sense.

How about `vite-plugin-ssr` *cumulates* all `addContextProps()`? You'd then simply add your global data to `_default.page.server.js`'s `addContextProps()`, and add additional data to pages by using the page's `.page.server.js`'s `addContextProps()`. Basically, `vite-plugin-ssr` wouldn't override `addContextProps()` anymore.

If overriding is wanted, you'd be able to `export skipDefaultAddContextProps = true` in the page's `.page.server.js`.

Thoughts?

**Comment by gryphonmyers** [comment]
@brillout personally I think two separate hooks is less confusing. Then we don't have to change the override rules, and with clear naming we can convey that one is for initializing and the other is for adding on after it has been initialized.

Another option if you wanted to go the override route rather than separate hooks, would be to expose some form of super reference. I know these aren't classes, but if you had a reference to the overridden function you could choose to call it inside the overriding one. I would find that more straightforward and powerful than manipulating the override rules with an exported bool.

**Comment by brillout** [comment]
> personally I think two separate hooks is less confusing

I agree, I'll think about it.

> if you had a reference to the overridden function you could choose to call it 

Yes I thought about this as well. The good thing is that you'd have more control over the order of calls, but the bad thing is that you'd always have to think about calling the overriden function.

E.g. one question I'm asking myself is whether the default/global function should be called after or before the page specific one?

**Comment by gryphonmyers** [comment]
> Yes I thought about this as well. The good thing is that you'd have more control over the order of calls, but the bad thing is that you'd always have to think about calling the overriden function.

That does sound less than ideal.

> E.g. one question I'm asking myself is whether the default/global function should be called after or before the page specific one?

In my mind the new hook would occur as early as possible, like here: https://github.com/brillout/vite-plugin-ssr/blob/master/src/renderPage.node.ts#L56 or here https://github.com/brillout/vite-plugin-ssr/blob/master/src/renderPage.node.ts#L76 if we want it to have all the parsed url information.

**Comment by brillout** [comment]
I like the hook thing more and more, I'm still thinking about it.

**Comment by brillout** [comment]
The whole idea of *stage* hooks is to have different hooks for different *stages* of the render process.

The `_default.page.*` files do *not* apply gobaly: as explained in https://github.com/brillout/vite-plugin-ssr#_defaultpage, there can be several `_default`'s that apply to different pages.

This means we cannot apply any hook before the routing determined what page should be rendered.

This means all hooks need to be called after routing. This leaves us with two sages: `onBeforeRender` and `onAfterRender`. So far, I don't see use cases for `onAfterRender` that cannot be already achieved today.

We could still create a `onInitializeContext` hook, but it would be called at the same time than `onBeforeRender`. This would be confusing and defeats the whole purpose of having different hooks for different stages.

That's my thinking so far.

**Comment by gryphonmyers** [comment]
That all makes sense. I had the same thought about "it seems like these two hooks will get called around the same time" which makes it seem odd for there to be two of them. As I've mentioned before though, the real objective here is to cut down on repetition when utilizing common context. I'm not necessarily attached to the `onInitializeContext` idea, but I think there is a significant usability problem here that needs to be addressed, and it's the most elegant solution I can think of at the moment. 

What about a function called `createContext` that would default to creating an empty object if unspecified?

**Comment by brillout** [comment]
> What about a function called `createContext` that would default to creating an empty object if unspecified?

Where would we define it? AFAICT we would need to spec an extra file `_createContext.page.js`.

<br/>

I'm also thinking about the following: when `addPageContext` is defined in *both* `_default.page.server.js` and the page's `.page.server.js` then we throw following error:

> [Wrong Usage] When defining the `addPageContext()` hook in /path/to/_default.page.server.js as well as in /path/to/product.page.server.js you need to either (a) `export const skipDefaultAddPageContext = true` in /path/to/product.page.server.js, or (b) call `pageContext.addPageContext(pageContext)` in your `addPageContext()` hook in /path/to/_default.page.server.js, more infos at https://github.com/brillout/vite-plugin-ssr#pagecontext

```js
// _default.page.server.js

export { addPageContext }

async function addPageContext({ pageContext }) {
  // ... (e.g. what you would put in `createContext`/`onInitializeContext`)

  // We call the `addPageContext()` hook of `product.page.server.js`
  const newPageContext = await pageContext.addPageContext(pageContext)

  // ...
}
```

**Comment by gryphonmyers** [comment]
> > What about a function called `createContext` that would default to creating an empty object if unspecified?
> 
> Where would we define it? AFAICT we would need to spec an extra file `_createContext.page.js`.
> 
> 
> I'm also thinking about the following: when `addPageContext` is defined in _both_ `_default.page.server.js` and the page's `.page.server.js` then we throw following error:
> 
> > [Wrong Usage] When defining the `addPageContext()` hook in /path/to/_default.page.server.js as well as in /path/to/product.page.server.js you need to either (a) `export const skipDefaultAddPageContext = true` in /path/to/product.page.server.js, or (b) call `pageContext.addPageContext(pageContext)` in your `addPageContext()` hook in /path/to/_default.page.server.js, more infos at https://github.com/brillout/vite-plugin-ssr#pagecontext
> 
> ```js
> // _default.page.server.js
> 
> export { addPageContext }
> 
> async function addPageContext({ pageContext }) {
>   // ... (e.g. what you would put in `createContext`/`onInitializeContext`)
> 
>   // We call the `addPageContext()` hook of `product.page.server.js`
>   const newPageContext = await pageContext.addPageContext(pageContext)
> 
>   // ...
> }
> ```

Interesting, so it's like calling super, but the default file is extending each page file rather than being extended by it. This gets around the problem we discussed earlier of having to call super in every page file. Is this architecturally feasible?

One thought - what if rather than `pageContext.addPageContext` the default file just got a reference to the "parent", e.g.

```javascript
// _default.page.server.js

export { addPageContext }

async function addPageContext({ pageContext, parentPage }) {
  // ... (e.g. what you would put in `createContext`/`onInitializeContext`)

  // We call the `addPageContext()` hook of `product.page.server.js`
  const newPageContext = await parentPage.addPageContext(pageContext)

  // ...
}
```

**Comment by brillout** [comment]
Neat idea.

How about `pageHooks.addPageContext`?

Note that everything is now defined on `pageContext`, see https://github.com/brillout/vite-plugin-ssr/blob/master/CHANGELOG.md#010-beta46-2021-05-28

So it would be `pageContext.pageHooks.addPageContext`. Cubersome to write but explicit & clear.

**Comment by gryphonmyers** [comment]
> Neat idea.
> 
> How about `pageHooks.addPageContext`?
> 
> Note that everything is now defined on `pageContext`, see https://github.com/brillout/vite-plugin-ssr/blob/master/CHANGELOG.md#010-beta46-2021-05-28
> 
> So it would be `pageContext.pageHooks.addPageContext`. Cubersome to write but explicit & clear.

Ah I see, I had missed that change. You've been busy :)

Thinking more about it, I'm not sure the inversion of the parent -> child relationship would be intuitive for people to understand. One would likely assume the default file is the parent, and the individual page files are extending it. Maybe we should just focus on making the process for calling "super" as succinct as possible so the burden isn't too great in each child page? We must bear in mind that this is also an entirely opt-in feature - one could simply ignore or not specify `addPageContext` in the parent, which would achieve the same functionality that we have currently. Maybe it's not too high a price to pay to have to do something like the following if global data is needed?

```javascript
// _default.page.server.js

export { addPageContext }

async function addPageContext(pageContext) {
 return Object.assign(pageContext, {
   foo: 'bar'
  });
}

// product.page.server.js

export { addPageContext }

async function addPageContext(pageContext) {
  return Object.assign(await pageContext.parentHooks.addPageContext(pageContext), {
    bar: 'foo'
  });
}
```

**Comment by gryphonmyers** [comment]
BTW I could see this feature playing very nicely with https://github.com/brillout/vite-plugin-ssr/issues/72, as it would allow you to granularly group common context into a hierarchy of page files suited to particular needs. For example if you had multiple pages that dealt with "news", you could create a `_news` parent that itself references `_default`, then have all "news" pages point to `_news`. Then in the `addPageContext` hook of each child page file, it can get not only global common context, but news-specific common context

**Comment by brillout** [comment]
If you look at the list of built-in properties defined on `pageContext` (https://github.com/brillout/vite-plugin-ssr#pagecontext) you'll see that `pageContext` is essentially a list of properties about the page being rendered. That's why I find it natural to make the *page*'s hooks available as `pageContext.pageHooks`.

The `_default` hooks are about orchestrating the overall behavior of your app; it makes sense to have the control flow defined in the `_default` hooks.

```js
// _default.page.server.js

export async function addPageContext(pageContext) {
  // Do all the stuff common to all pages
  // ...

  // We have exact control about how we glue the default `addPageContext`
  // with the page's `addPageContext`.
  const pageContextAddenndum = await pageContext.pageHooks.addPageContext(pageContext)

  // ...
}
```

That's full control for the `_default` hooks. The `_default` hook is the place where we want to be able to be clever to make defining new pages as easy and as dumb as possible. That's, actually, the whole purpose of defining _default hooks: make creating new pages easier.

About hierarchical defaults: I'm thinking maybe we don't need it. The main idea here is that even a large-scale enterprise app would have a max of 2-3 different defaults.

```js
// hooks/addTranslations.js

export { addTranslations }
```

```js
// landing-page/_default.page.server.js

import { addTranslations } from '../hooks/addTranslations'

export async function addPageContext(pageContext) {
  const translations = addTranslations(pageContext)
  const landingPageStuff = /*...*/;
  return { translations, landingPageStuff }
}
```

```js
// admin-panel/_default.page.server.js

import { addTranslations } from '../hooks/addTranslations'

export async function addPageContext(pageContext) {
  const translations = addTranslations(pageContext)
  const adminPanelStuff = /*...*/;
  return { translations, adminPanelStuff }
}
```

That's how you could share common `pageContext` between defaults. Since there would be only a max of 2-3 different defaults, that's I would argue a viable situation.

**Comment by brillout** [comment]
Closing in favor of https://github.com/brillout/vite-plugin-ssr/issues/164
